File: src/voice_agent/tts.py
"""Text-to-Speech using OpenAI TTS API."""

import asyncio
import os
from typing import AsyncIterator

from openai import AsyncOpenAI


# OpenAI TTS outputs 24kHz PCM by default with pcm format
TTS_SAMPLE_RATE = 24000


class TTSClient:
    """Client for OpenAI TTS API - much cheaper than realtime for tool outputs."""

    def __init__(
        self,
        api_key: str | None = None,
        model: str = "gpt-4o-mini-tts",  # tts-1 is fast and cheap, tts-1-hd for quality
        voice: str = "echo",
        instructions: str = "Speak in a cheerful and positive tone.",
    ):
        """Initialize TTS client.

        Args:
            api_key: OpenAI API key (uses env var if not provided)
            model: TTS model (tts-1 or tts-1-hd)
            voice: Voice to use (alloy, echo, fable, onyx, nova, shimmer)
            instructions: Global instructions for speech style/tone
        """
        self.api_key = api_key or os.environ.get("OPENAI_API_KEY")
        self.model = model
        self.voice = voice
        self.instructions = instructions
        self._client: AsyncOpenAI | None = None

    @property
    def client(self) -> AsyncOpenAI:
        """Get or create the OpenAI client."""
        if self._client is None:
            self._client = AsyncOpenAI(api_key=self.api_key)
        return self._client

    async def synthesize(self, text: str) -> bytes:
        """Synthesize text to speech.

        Args:
            text: Text to synthesize

        Returns:
            PCM16 audio bytes at 24kHz
        """
        response = await self.client.audio.speech.create(
            model=self.model,
            voice=self.voice,
            input=text,
            instructions=self.instructions,
            response_format="pcm",  # Raw PCM16 at 24kHz
        )

        # Read all content
        return response.content

    async def synthesize_stream(self, text: str) -> AsyncIterator[bytes]:
        """Synthesize text to speech with streaming.

        Args:
            text: Text to synthesize

        Yields:
            PCM16 audio chunks at 24kHz
        """
        async with self.client.audio.speech.with_streaming_response.create(
            model=self.model,
            voice=self.voice,
            input=text,
            instructions=self.instructions,
            response_format="pcm",
        ) as response:
            async for chunk in response.iter_bytes(chunk_size=4096):
                yield chunk

File: packages/voice-agent/src/audio/tts.ts
/**
 * Text-to-Speech - Generate speech using OpenAI TTS API
 *
 * Streams PCM audio directly to PipeWire for playback.
 */

import OpenAI from 'openai';
import { spawn } from 'child_process';
import { config } from '../config.js';

const openai = new OpenAI({
  apiKey: config.openai.apiKey,
});

const TTS_SAMPLE_RATE = 24000; // OpenAI TTS outputs 24kHz

export interface TTSOptions {
  voice?: 'alloy' | 'ash' | 'coral' | 'echo' | 'fable' | 'onyx' | 'nova' | 'shimmer';
  speed?: number;
}

/**
 * Speak text using OpenAI TTS API, streaming PCM directly to PipeWire.
 */
export async function speak(text: string, options: TTSOptions = {}): Promise<void> {
  const voice = options.voice ?? 'echo';
  const speed = options.speed ?? 1.0;

  console.log(`[TTS] Speaking: "${text.substring(0, 50)}${text.length > 50 ? '...' : ''}"`);

  // Start pw-cat for PCM playback
  const player = spawn('pw-cat', [
    '--playback',
    '--raw',
    '--channels', '1',
    '--rate', String(TTS_SAMPLE_RATE),
    '--format', 's16',
    '-',
  ], {
    stdio: ['pipe', 'pipe', 'pipe'],
  });

  let playerError: Error | null = null;

  player.stderr?.on('data', (data: Buffer) => {
    const msg = data.toString().trim();
    if (msg && !msg.includes('opening stream')) {
      console.error('[TTS] pw-cat:', msg);
    }
  });

  player.on('error', (err) => {
    playerError = err;
  });

  const playerDone = new Promise<void>((resolve, reject) => {
    player.on('exit', (code) => {
      if (code === 0 || code === null) {
        resolve();
      } else if (playerError) {
        reject(playerError);
      } else {
        reject(new Error(`pw-cat exited with code ${code}`));
      }
    });
  });

  try {
    // Request PCM audio from OpenAI
    const response = await openai.audio.speech.create({
      model: 'gpt-4o-mini-tts',
      voice,
      input: text,
      speed,
      response_format: 'pcm', // 24kHz 16-bit mono PCM
    });

    // Stream the audio data to pw-cat
    const arrayBuffer = await response.arrayBuffer();
    const buffer = Buffer.from(arrayBuffer);

    player.stdin?.write(buffer);
    player.stdin?.end();

    // Wait for playback to complete
    await playerDone;
    console.log('[TTS] Done');
  } catch (error) {
    // Kill the player on error
    player.kill();
    console.error('[TTS] Error:', error);
    throw error;
  }
}
